### ISSUE TITLE
Auto-fixing not working with PRs from forks

### ISSUE THREAD

**Issue by samuelmeuli** [samuelmeuli/lint-action#13]
Auto-fixing works as expected for `pull_request` events when the code is on a branch in the same repository.

Unfortunately, it currently doesn't work when PRs are created from a fork: The `pull_request` event is not triggered on the fork, but on the main repository (which is what `origin` points to). Checking out and pushing to the fork branch therefore fails because it's not available to the action.

See https://github.com/facebook/jest/pull/9367

**Comment by samuelmeuli** [comment]
It seems like this is impossible with GitHub's current token scopes.

The action needs permissions for two operations:

- Permission to create annotations with GitHub's [Check Runs API](https://developer.github.com/v3/checks/runs) to display the user's linting errors
- Permission to push changes with Git (if `auto_fix: true`)

If the action is triggered on changes to the main repository, it will work because the `GITHUB_TOKEN` has permissions to modify that repository. On forks, however, this becomes tricky to impossible:

| | **`push`** event | **`pull_request`** event |
| ------------- | ------------- | ----- |
| **commit to main repo**      | Works | Works |
| **commit to fork**     | Only runs after the user manually activates GitHub Actions on the fork. Even then, annotations will not display on the PR (only on the fork). | No permission to push auto-fix changes. No permission to create annotations for linting errors. |

Note that the `pull_request` event always runs on the main repository, even for pushes to the fork. The token scopes are defined accordingly.

If anybody is aware of a workaround, I'd be very happy to hear it :)

_Related: https://github.com/actions/checkout/issues/124#issuecomment-580018424_

**Comment by samuelmeuli** [comment]
All I can do for now is document these limitations in the README and provide more helpful error messages.

If you'd also like to see these permissions changed, please [submit feedback to GitHub](https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions).

**Comment by MarvinT** [comment]
Would it be possible to catch the 403 error when the PR is coming from a fork so it doesn't mark the action as having failed?

**Comment by ocean90** [comment]
GitHub introduced a new `pull_request_target ` event which might help here, see https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/#improvements-for-public-repository-forks

**Comment by 1212gmartinez** [comment]
@samuelmeuli Thanks for making this easy to use library!

Have you had any chance to consider supporting github's `pull_request_target`.
It would be great to have the fork support available.

**Comment by ocean90** [comment]
@1212gmartinez There's a WIP feature branch for this which you can try: https://github.com/wearerequired/lint-action/tree/feature/pull_request_target

You should be able to use it with `wearerequired/lint-action@feature/pull_request_target`.

**Comment by 1212gmartinez** [comment]
Awesome, thanks!

**Comment by moodysalem** [comment]
Seems not to be working for us
```
fatal: Cannot force update the current branch.
Error: Command failed: git branch --force master --track fork/master
fatal: Cannot force update the current branch.
```
https://github.com/Uniswap/uniswap-interface/pull/1121/checks?check_run_id=1164841104

**Comment by moodysalem** [comment]
possibly related?
https://stackoverflow.com/questions/29559053/force-update-the-current-branch-how

issue is with `git branch --force master --track fork/master`

is it because the fork branches are named master?

in another case the linter passes but the check status is never reported

https://github.com/Uniswap/uniswap-interface/pull/1060
![image](https://user-images.githubusercontent.com/7897876/96035609-e2b62280-0e28-11eb-943d-e547aafffcb0.png)

**Comment by yuekui** [comment]
Instead of using `pull_request_target`, we could check repo's owner to skip lint action:
```
    - name: Run Lint
      uses: wearerequired/lint-action@v1
      if: ${{ github.event_name != 'pull_request' || github.event.pull_request.head.repo.owner.login == github.repository_owner }}
      with:
        github_token: ${{ secrets.github_token }}
        black: true
        flake8: true
        auto_fix: true
```
So that lint action would be run only on `push` or `pull_request` events within the same repository.

**Comment by jnm2** [comment]
If you write workflow commands to stdout instead of making web API calls with the token, the authentication issue goes away and the performance may even be better.

Example from docs page (https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message):

```
echo "::warning file=app.js,line=1,col=5::Missing semicolon"
```

A GitHub action using this approach: https://github.com/cschleiden/jest-github-actions-reporter

**Comment by alexbbt** [comment]
`pull_request_target` worked very well for me. See my PR here for an example: https://github.com/join-monster/join-monster/pull/449. Read this before using this new event though, as it has security implications: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

**Comment by Thatkookooguy** [comment]
Hey! Was this solved? I see that the `pull_request_target` branch has been merge and `pull_request_target` is being referenced in code.

But still when I try to use `pull_request_target` with this aciton and a fork opens a PR, I get the 404 error.

see [this dependabot PR for example](https://github.com/Kibibit/achievibit/pull/368)

**Comment by ocean90** [comment]
@Thatkookooguy Pull requests by Dependabot don't have access to secrets anymore, see https://github.com/dependabot/dependabot-core/issues/3253#issuecomment-852541544.
You should be able to use the the default `GITHUB_TOKEN` secret in combination with `pull_request_target`. Consider also [limiting the default `permissions`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions). For a working example see [test-action.yml](https://github.com/wearerequired/lint-action/blob/master/.github/workflows/test-action.yml).

**Comment by jnm2** [comment]
Please consider using workflow commands (stdout) so that tokens are not needed and pull_request_target is not needed, as mentioned above (https://github.com/wearerequired/lint-action/issues/13#issuecomment-811518586).

**Comment by Thatkookooguy** [comment]
@ocean90 thanks for the tip! will definitely check it out.

**Comment by jnm2** [comment]
@ocean90 I'm confused. Why did you downvote?

**Comment by moodysalem** [comment]
> Please consider using workflow commands (stdout) so that tokens are not needed and pull_request_target is not needed, as mentioned above ([#13 (comment)](https://github.com/wearerequired/lint-action/issues/13#issuecomment-811518586)).

This particular change seems like it should not be blocked, but won't help with `auto_fix: true`

@samuelmeuli Will you take a PR? Seems like this piece of code needs changing:
